[BUG] NullReferenceException pressing Android 'back' button when (IDialogService) Dialog is displayed [MAUI/9.0.271-pre]

### Description

If you display a dialog using IDialogService and then tap the Android back button whilst the dialog is displayed, the application crashes with a NullReferenceException.

Repo: 
[PrismIssue.zip](https://github.com/PrismLibrary/Prism/files/13517989/PrismIssue.zip)


### Steps to Reproduce

Unzip the attached repo.
Open in Visual Studio 2022 (I am using 17.8.2)
Build and deploy to an Android device that has a back button.
Click the button on the first screen to open the dialog.
Press the Android back button

### Platform with bug

.NET MAUI

### Affected platforms

Android, I was *not* able test on other platforms

### Did you find any workaround?

The only workaround I have to date is to disable the back button in the MainActivity of the Android application - this works for the specific project I am working on but will be a pain for other up-coming projects.

### Relevant log output

```shell
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object.
[mono-rt]    at Prism.Ioc.IContainerProviderExtensions.Resolve[INavigationService](IContainerProvider provider) in /_/src/Prism.Container.Abstractions/IContainerProviderExtensions.cs:line 16
[mono-rt]    at Prism.Navigation.Xaml.Navigation.GetNavigationService(Page page) in D:\a\Prism\Prism\src\Maui\Prism.Maui\Navigation\Xaml\Navigation.cs:line 117
[mono-rt]    at Prism.Navigation.PrismWindow.PrismWindow_ModalPopping(Object sender, ModalPoppingEventArgs e) in D:\a\Prism\Prism\src\Maui\Prism.Maui\Navigation\PrismWindow.cs:line 82
[mono-rt]    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
[mono-rt]    at Android.App.SyncContext.<>c__DisplayClass2_0.<Post>b__0() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:line 36
[mono-rt]    at Java.Lang.Thread.RunnableImplementor.Run() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:line 36
[mono-rt]    at Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Java.Lang.IRunnable.cs:line 84
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 26
```

--- a/src/Maui/Prism.Maui/Dialogs/DialogService.cs
+++ b/src/Maui/Prism.Maui/Dialogs/DialogService.cs
@@ -1,4 +1,4 @@
-﻿using Prism.Commands;
+using Prism.Commands;
 using Prism.Common;
 using Prism.Ioc;
 using Prism.Navigation;
@@ -29,6 +29,7 @@ public sealed class DialogService : IDialogService
     /// <inheritdoc/>
     public void ShowDialog(string name, IDialogParameters parameters, DialogCallback callback)
     {
+        IDialogContainer? dialogModal = null;
         try
         {
             parameters = UriParsingHelper.GetSegmentParameters(name, parameters ?? new DialogParameters());
@@ -39,16 +40,19 @@ public sealed class DialogService : IDialogService
             var view = registry.CreateView(_container, UriParsingHelper.GetSegmentName(name)) as View;
 
             var currentPage = _pageAccessor.Page;
-            var dialogModal = _container.Resolve<IDialogContainer>();
+            dialogModal = _container.Resolve<IDialogContainer>();
+            IDialogContainer.DialogStack.Add(dialogModal);
             var dialogAware = GetDialogController(view);
 
             async Task DialogAware_RequestClose(IDialogResult outResult)
             {
+                bool didCloseDialog = true;
                 try
                 {
                     var result = await CloseDialogAsync(outResult ?? new DialogResult(), currentPage, dialogModal);
                     if (result.Exception is DialogException de && de.Message == DialogException.CanCloseIsFalse)
                     {
+                        didCloseDialog = false;
                         return;
                     }
 
@@ -57,6 +61,12 @@ public sealed class DialogService : IDialogService
                 }
                 catch (DialogException dex)
                 {
+                    if (dex.Message == DialogException.CanCloseIsFalse)
+                    {
+                        didCloseDialog = false;
+                        return;
+                    }
+
                     var result = new DialogResult
                     {
                         Exception = dex,
@@ -72,6 +82,13 @@ public sealed class DialogService : IDialogService
                 catch (Exception ex)
                 {
                     await InvokeError(callback, ex, parameters);
+                }
+                finally
+                {
+                    if (didCloseDialog && dialogModal is not null)
+                    {
+                        IDialogContainer.DialogStack.Remove(dialogModal);
+                    }
                 }
             }
 

--- a/src/Maui/Prism.Maui/Dialogs/IDialogContainer.cs
+++ b/src/Maui/Prism.Maui/Dialogs/IDialogContainer.cs
@@ -1,11 +1,46 @@
-﻿using System.Windows.Input;
+using System.Windows.Input;
 
+#nullable enable
 namespace Prism.Dialogs;
 
+/// <summary>
+/// Interface representing a container for managing dialogs.
+/// </summary>
 public interface IDialogContainer
 {
+    /// <summary>
+    /// Gets a stack of currently active dialog containers.
+    /// </summary>
+    /// <remarks>
+    /// This property provides access to the list of dialogs currently displayed, typically in the order they were presented.
+    /// </remarks>
+    static IList<IDialogContainer> DialogStack { get; } = [];
+
+    /// <summary>
+    /// Gets the view associated with the currently displayed dialog.
+    /// </summary>
     View DialogView { get; }
+
+    /// <summary>
+    /// Gets a command that can be used to dismiss the currently displayed dialog.
+    /// </summary>
     ICommand Dismiss { get; }
+
+    /// <summary>
+    /// Configures the layout and behavior of the dialog.
+    /// </summary>
+    /// <param name="currentPage">The page on which the dialog is being displayed.</param>
+    /// <param name="dialogView">The view representing the dialog content.</param>
+    /// <param name="hideOnBackgroundTapped">True if the dialog should close when the background is tapped, false otherwise.</param>
+    /// <param name="dismissCommand">The command to execute when the dialog is dismissed.</param>
+    /// <param name="parameters">Optional parameters to pass to the dialog.</param>
+    /// <returns>A task representing the completion of the configuration process.</returns>
     Task ConfigureLayout(Page currentPage, View dialogView, bool hideOnBackgroundTapped, ICommand dismissCommand, IDialogParameters parameters);
+
+    /// <summary>
+    /// Performs the internal logic to display or dismiss the dialog.
+    /// </summary>
+    /// <param name="currentPage">The page on which the dialog is being displayed.</param>
+    /// <returns>A task representing the completion of the pop operation.</returns>
     Task DoPop(Page currentPage);
 }

--- a/src/Maui/Prism.Maui/Dialogs/RelativeContentSizeConverter.cs
+++ b/src/Maui/Prism.Maui/Dialogs/RelativeContentSizeConverter.cs
@@ -1,4 +1,4 @@
-﻿using System.Globalization;
+using System.Globalization;
 
 namespace Prism.Dialogs;
 
@@ -12,7 +12,7 @@ internal class RelativeContentSizeConverter : IValueConverter
         {
             if (value == 0)
             {
-                relativeSize = 1;
+                relativeSize = -1;
             }
             else if (value > 1)
             {

--- a/src/Maui/Prism.Maui/Navigation/PrismWindow.cs
+++ b/src/Maui/Prism.Maui/Navigation/PrismWindow.cs
@@ -1,4 +1,4 @@
-﻿using System.ComponentModel;
+using System.ComponentModel;
 using Prism.AppModel;
 using Prism.Common;
 using Prism.Dialogs;
@@ -79,8 +79,17 @@ internal class PrismWindow : Window
         if (PageNavigationService.NavigationSource == PageNavigationSource.Device)
         {
             e.Cancel = true;
-            var navService = Xaml.Navigation.GetNavigationService(e.Modal);
-            await navService.GoBackAsync();
+            var dialogModal = IDialogContainer.DialogStack.LastOrDefault();
+            if (dialogModal is not null)
+            {
+                if (dialogModal.Dismiss.CanExecute(null))
+                    dialogModal.Dismiss.Execute(null);
+            }
+            else
+            {
+                var navService = Xaml.Navigation.GetNavigationService(e.Modal);
+                await navService.GoBackAsync();
+            }
         }
     }
 

--- a/src/Maui/Prism.Maui/PrismAppBuilder.cs
+++ b/src/Maui/Prism.Maui/PrismAppBuilder.cs
@@ -46,6 +46,9 @@ public sealed class PrismAppBuilder
             return $"View is not a BindableObject: '{view.GetType().FullName}";
         });
 
+        // Ensure that the DialogStack is cleared when the Application is started.
+        // This is primarily to help with Unit Tests
+        IDialogContainer.DialogStack.Clear();
         MauiBuilder = builder;
         MauiBuilder.ConfigureContainer(new PrismServiceProviderFactory(RegistrationCallback));
         MauiBuilder.ConfigureLifecycleEvents(lifecycle =>
@@ -55,6 +58,15 @@ public sealed class PrismAppBuilder
             {
                 android.OnBackPressed(activity =>
                 {
+                    var dialogModal = IDialogContainer.DialogStack.LastOrDefault();
+                    if (dialogModal is not null)
+                    {
+                        if (dialogModal.Dismiss.CanExecute(null))
+                            dialogModal.Dismiss.Execute(null);
+
+                        return true;
+                    }
+
                     var root = ContainerLocator.Container;
                     if (root is null)
                         return false;

--- a/src/Maui/Prism.Maui/PrismAppBuilder.cs
+++ b/src/Maui/Prism.Maui/PrismAppBuilder.cs
@@ -72,14 +72,14 @@ public sealed class PrismAppBuilder
                         return false;
 
                     var app = root.Resolve<IApplication>();
-                    var windows = app.Windows.OfType<PrismWindow>();
-                    if (!windows.Any(x => x.IsActive))
+                    var window = app.Windows.OfType<PrismWindow>()
+                        .FirstOrDefault(x => x.IsActive);
+
+                    if (window is null)
                         return false;
 
-                    var window = windows.First(x => x.IsActive);
-                    if(window.IsRootPage && app is Application application)
+                    if(window.IsRootPage)
                     {
-                        application.Quit();
                         return false;
                     }
 

